src/auxsolvers/vector_gridfunction_cross_product_aux.hpp

```diff
@@ -34,8 +34,8 @@ class VectorGridFunctionCrossProductCoefficient : public mfem::VectorCoefficient
 class VectorGridFunctionCrossProductAux : public VectorCoefficientAux
 {
 private:
-  mfem::ParGridFunction * _u_gf{nullptr};
-  mfem::ParGridFunction * _v_gf{nullptr};
+  std::shared_ptr<mfem::ParGridFunction> _u_gf{nullptr};
```

Comment:
Wider philosophical question here - do we actually want auxsolvers/boundary conditions/coefficients etc. to share ownership, given that should be managed by their respective containers?

Clearly `hephaestus::NamedFieldsMap` should store (at least) shared_ptrs (I think unique_ptrs would also be possible for hephaestus in isolation, but we know that'll cause headaches alongside MOOSE/Apollo). And there should be no need for naked 'new' with the new Register method calling `make_shared`. But individual auxsolvers/bcs/coefficients etc. shouldn't need to share memory management of those objects.

Thoughts?

Reply:
Ideally, yes. However, I would make the point that shared pointers are much better for casting. Sometimes we're creating a shared pointer to a derived class instance shared pointer and passing this as an argument to an initializer which takes a shared pointer to the base class.

Reply:
An example, CoupledCoefficient has a shared pointer member variable _gf which it sets by obtaining a shared pointer from a NamedFieldsMap so it will share ownership with the NamedFieldMap.

Reply:
Shared pointers were the quickest and simplest solution for quickly fixing a number of memory leaks without making major changes to the classes. But I think in the future, some of the classes should certainly be refactored a little and we should attempt to replace shared pointers with unique pointers were possible. In these cases, it may be better to rewrite the initializer so that we don't create shared/unique pointers to objects outside the class and then pass them in.

Reply:
Ah sorry - my initial comment wasn't clear. I meant, do we need to use smart pointers at all for individual auxsolvers/coefficients/bcs/kernels, since they shouldn't own them anyway?